File: app/scala/part3/src/main/scala/CompositePattern.scala
/**
 * 第7章: Composite パターン
 * 
 * Composite パターンは、オブジェクトをツリー構造で構成し、
 * 個々のオブジェクトとオブジェクトの集合を同じように扱うことができるようにするパターンです。
 */
package compositepattern

// ============================================
// 1. Shape の例 - 図形の Composite パターン
// ============================================

/** 2D座標 */
case class Point(x: Double, y: Double):
  def +(other: Point): Point = Point(x + other.x, y + other.y)
  def *(factor: Double): Point = Point(x * factor, y * factor)

/** 図形の共通インターフェース（sealed trait） */
sealed trait Shape:
  /** 図形を移動する */
  def translate(dx: Double, dy: Double): Shape
  
  /** 図形を拡大/縮小する */
  def scale(factor: Double): Shape
  
  /** 面積を計算する */
  def area: Double
  
  /** バウンディングボックスを取得 */
  def boundingBox: BoundingBox

/** バウンディングボックス */
case class BoundingBox(topLeft: Point, bottomRight: Point):
  def width: Double = bottomRight.x - topLeft.x
  def height: Double = bottomRight.y - topLeft.y
  def area: Double = width * height
  
  /** 二つのバウンディングボックスを結合 */
  def union(other: BoundingBox): BoundingBox =
    BoundingBox(
      Point(math.min(topLeft.x, other.topLeft.x), math.min(topLeft.y, other.topLeft.y)),
      Point(math.max(bottomRight.x, other.bottomRight.x), math.max(bottomRight.y, other.bottomRight.y))
    )

/** 円（Leaf） */
case class Circle(center: Point, radius: Double) extends Shape:
  def translate(dx: Double, dy: Double): Circle =
    copy(center = Point(center.x + dx, center.y + dy))
  
  def scale(factor: Double): Circle =
    copy(radius = radius * factor)
  
  def area: Double = math.Pi * radius * radius
  
  def boundingBox: BoundingBox =
    BoundingBox(
      Point(center.x - radius, center.y - radius),
      Point(center.x + radius, center.y + radius)
    )

/** 正方形（Leaf） */
case class Square(topLeft: Point, side: Double) extends Shape:
  def translate(dx: Double, dy: Double): Square =
    copy(topLeft = Point(topLeft.x + dx, topLeft.y + dy))
  
  def scale(factor: Double): Square =
    copy(side = side * factor)
  
  def area: Double = side * side
  
  def boundingBox: BoundingBox =
    BoundingBox(topLeft, Point(topLeft.x + side, topLeft.y + side))

/** 長方形（Leaf） */
case class Rectangle(topLeft: Point, width: Double, height: Double) extends Shape:
  def translate(dx: Double, dy: Double): Rectangle =
    copy(topLeft = Point(topLeft.x + dx, topLeft.y + dy))
  
  def scale(factor: Double): Rectangle =
    copy(width = width * factor, height = height * factor)
  
  def area: Double = width * height
  
  def boundingBox: BoundingBox =
    BoundingBox(topLeft, Point(topLeft.x + width, topLeft.y + height))

/** 三角形（Leaf） */
case class Triangle(p1: Point, p2: Point, p3: Point) extends Shape:
  def translate(dx: Double, dy: Double): Triangle =
    Triangle(
      Point(p1.x + dx, p1.y + dy),
      Point(p2.x + dx, p2.y + dy),
      Point(p3.x + dx, p3.y + dy)
    )
  
  def scale(factor: Double): Triangle =
    // 中心を基準にスケール
    val cx = (p1.x + p2.x + p3.x) / 3
    val cy = (p1.y + p2.y + p3.y) / 3
    Triangle(
      Point(cx + (p1.x - cx) * factor, cy + (p1.y - cy) * factor),
      Point(cx + (p2.x - cx) * factor, cy + (p2.y - cy) * factor),
      Point(cx + (p3.x - cx) * factor, cy + (p3.y - cy) * factor)
    )
  
  def area: Double =
    math.abs((p2.x - p1.x) * (p3.y - p1.y) - (p3.x - p1.x) * (p2.y - p1.y)) / 2
  
  def boundingBox: BoundingBox =
    BoundingBox(
      Point(List(p1.x, p2.x, p3.x).min, List(p1.y, p2.y, p3.y).min),
      Point(List(p1.x, p2.x, p3.x).max, List(p1.y, p2.y, p3.y).max)
    )

/** 複合図形（Composite） */
case class CompositeShape(shapes: List[Shape] = Nil) extends Shape:
  def add(shape: Shape): CompositeShape = copy(shapes = shapes :+ shape)
  
  def remove(shape: Shape): CompositeShape = copy(shapes = shapes.filterNot(_ == shape))
  
  def translate(dx: Double, dy: Double): CompositeShape =
    copy(shapes = shapes.map(_.translate(dx, dy)))
  
  def scale(factor: Double): CompositeShape =
    copy(shapes = shapes.map(_.scale(factor)))
  
  def area: Double = shapes.map(_.area).sum
  
  def boundingBox: BoundingBox =
    shapes match
      case Nil => BoundingBox(Point(0, 0), Point(0, 0))
      case head :: tail => tail.foldLeft(head.boundingBox)((acc, s) => acc.union(s.boundingBox))
  
  def count: Int = shapes.length
  
  def flatten: List[Shape] =
    shapes.flatMap {
      case cs: CompositeShape => cs.flatten
      case s => List(s)
    }

object CompositeShape:
  def apply(shapes: Shape*): CompositeShape = CompositeShape(shapes.toList)

// ============================================
// 2. Switchable の例 - スイッチの Composite パターン
// ============================================

/** スイッチの共通インターフェース */
sealed trait Switchable:
  def turnOn: Switchable
  def turnOff: Switchable
  def isOn: Boolean

/** 照明（Leaf） */
case class Light(on: Boolean = false, name: String = "Light") extends Switchable:
  def turnOn: Light = copy(on = true)
  def turnOff: Light = copy(on = false)
  def isOn: Boolean = on

/** 調光可能な照明（Leaf） */
case class DimmableLight(intensity: Int = 0, name: String = "DimmableLight") extends Switchable:
  def turnOn: DimmableLight = copy(intensity = 100)
  def turnOff: DimmableLight = copy(intensity = 0)
  def isOn: Boolean = intensity > 0
  
  def setIntensity(value: Int): DimmableLight =
    copy(intensity = math.max(0, math.min(100, value)))

/** 扇風機（Leaf） */
case class Fan(speed: Int = 0, name: String = "Fan") extends Switchable:
  def turnOn: Fan = copy(speed = 3) // 中速でオン
  def turnOff: Fan = copy(speed = 0)
  def isOn: Boolean = speed > 0
  
  def setSpeed(value: Int): Fan =
    copy(speed = math.max(0, math.min(5, value)))

/** 複合スイッチ（Composite） */
case class CompositeSwitchable(
  switchables: List[Switchable] = Nil,
  name: String = "Group"
) extends Switchable:
  
  def add(switchable: Switchable): CompositeSwitchable =
    copy(switchables = switchables :+ switchable)
  
  def remove(switchable: Switchable): CompositeSwitchable =
    copy(switchables = switchables.filterNot(_ == switchable))
  
  def turnOn: CompositeSwitchable =
    copy(switchables = switchables.map(_.turnOn))
  
  def turnOff: CompositeSwitchable =
    copy(switchables = switchables.map(_.turnOff))
  
  def isOn: Boolean = switchables.exists(_.isOn)
  
  def allOn: Boolean = switchables.nonEmpty && switchables.forall(_.isOn)
  
  def count: Int = switchables.length
  
  def flatten: List[Switchable] =
    switchables.flatMap {
      case cs: CompositeSwitchable => cs.flatten
      case s => List(s)
    }

object CompositeSwitchable:
  def apply(switchables: Switchable*): CompositeSwitchable =
    CompositeSwitchable(switchables.toList)

// ============================================
// 3. FileSystem の例 - ファイルシステムの Composite パターン
// ============================================

/** ファイルシステムエントリの共通インターフェース */
sealed trait FileSystemEntry:
  def name: String
  def size: Long
  def path: String
  def find(predicate: FileSystemEntry => Boolean): List[FileSystemEntry]

/** ファイル（Leaf） */
case class File(name: String, size: Long, parentPath: String = "") extends FileSystemEntry:
  def path: String = if parentPath.isEmpty then name else s"$parentPath/$name"
  
  def find(predicate: FileSystemEntry => Boolean): List[FileSystemEntry] =
    if predicate(this) then List(this) else Nil

/** ディレクトリ（Composite） */
case class Directory(
  name: String,
  children: List[FileSystemEntry] = Nil,
  parentPath: String = ""
) extends FileSystemEntry:
  
  def path: String = if parentPath.isEmpty then name else s"$parentPath/$name"
  
  def add(entry: FileSystemEntry): Directory =
    val newEntry = entry match
      case f: File => f.copy(parentPath = path)
      case d: Directory => d.copy(parentPath = path)
    copy(children = children :+ newEntry)
  
  def remove(entryName: String): Directory =
    copy(children = children.filterNot(_.name == entryName))
  
  def size: Long = children.map(_.size).sum
  
  def find(predicate: FileSystemEntry => Boolean): List[FileSystemEntry] =
    val selfMatch = if predicate(this) then List(this) else Nil
    val childMatches = children.flatMap(_.find(predicate))
    selfMatch ++ childMatches
  
  def fileCount: Int =
    children.foldLeft(0) { (acc, entry) =>
      entry match
        case _: File => acc + 1
        case d: Directory => acc + d.fileCount
    }
  
  def directoryCount: Int =
    children.foldLeft(0) { (acc, entry) =>
      entry match
        case _: File => acc
        case d: Directory => acc + 1 + d.directoryCount
    }
  
  def listAll: List[String] =
    children.flatMap {
      case f: File => List(f.path)
      case d: Directory => 
        val childDir = d.copy(parentPath = path)
        childDir.path :: childDir.listAll
    }

// ============================================
// 4. Menu の例 - メニューの Composite パターン
// ============================================

/** メニュー項目の共通インターフェース */
sealed trait MenuItem:
  def name: String
  def price: BigDecimal
  def isVegetarian: Boolean
  def description: String

/** 単品メニュー（Leaf） */
case class SingleItem(
  name: String,
  price: BigDecimal,
  isVegetarian: Boolean,
  description: String
) extends MenuItem

/** セットメニュー（Composite） */
case class SetMenu(
  name: String,
  items: List[MenuItem] = Nil,
  discount: BigDecimal = BigDecimal(0),
  description: String = ""
) extends MenuItem:
  
  def add(item: MenuItem): SetMenu = copy(items = items :+ item)
  
  def remove(itemName: String): SetMenu =
    copy(items = items.filterNot(_.name == itemName))
  
  /** セットメニューの価格（割引適用後） */
  def price: BigDecimal =
    val total = items.map(_.price).sum
    total - discount
  
  /** セット内のすべてがベジタリアンならtrue */
  def isVegetarian: Boolean = items.forall(_.isVegetarian)
  
  def itemCount: Int = items.length
  
  /** 元の合計価格 */
  def originalPrice: BigDecimal = items.map(_.price).sum
  
  /** 割引率 */
  def discountRate: Double =
    if originalPrice > 0 then (discount / originalPrice).toDouble * 100 else 0

// ============================================
// 5. Expression の例 - 数式の Composite パターン
// ============================================

/** 数式の共通インターフェース */
sealed trait Expression:
  def evaluate: Double
  def simplify: Expression
  def variables: Set[String]

/** 数値（Leaf） */
case class Number(value: Double) extends Expression:
  def evaluate: Double = value
  def simplify: Expression = this
  def variables: Set[String] = Set.empty
  override def toString: String = value.toString

/** 変数（Leaf） */
case class Variable(name: String, value: Option[Double] = None) extends Expression:
  def evaluate: Double = value.getOrElse(
    throw new IllegalStateException(s"Variable $name has no value")
  )
  def simplify: Expression = value.map(Number(_)).getOrElse(this)
  def variables: Set[String] = if value.isEmpty then Set(name) else Set.empty
  override def toString: String = name

/** 加算（Composite） */
case class Add(left: Expression, right: Expression) extends Expression:
  def evaluate: Double = left.evaluate + right.evaluate
  
  def simplify: Expression =
    (left.simplify, right.simplify) match
      case (Number(0), r) => r
      case (l, Number(0)) => l
      case (Number(a), Number(b)) => Number(a + b)
      case (l, r) => Add(l, r)
  
  def variables: Set[String] = left.variables ++ right.variables
  override def toString: String = s"($left + $right)"

/** 減算（Composite） */
case class Subtract(left: Expression, right: Expression) extends Expression:
  def evaluate: Double = left.evaluate - right.evaluate
  
  def simplify: Expression =
    (left.simplify, right.simplify) match
      case (l, Number(0)) => l
      case (Number(a), Number(b)) => Number(a - b)
      case (l, r) if l == r => Number(0)
      case (l, r) => Subtract(l, r)
  
  def variables: Set[String] = left.variables ++ right.variables
  override def toString: String = s"($left - $right)"

/** 乗算（Composite） */
case class Multiply(left: Expression, right: Expression) extends Expression:
  def evaluate: Double = left.evaluate * right.evaluate
  
  def simplify: Expression =
    (left.simplify, right.simplify) match
      case (Number(0), _) => Number(0)
      case (_, Number(0)) => Number(0)
      case (Number(1), r) => r
      case (l, Number(1)) => l
      case (Number(a), Number(b)) => Number(a * b)
      case (l, r) => Multiply(l, r)
  
  def variables: Set[String] = left.variables ++ right.variables
  override def toString: String = s"($left * $right)"

/** 除算（Composite） */
case class Divide(left: Expression, right: Expression) extends Expression:
  def evaluate: Double =
    val r = right.evaluate
    if r == 0 then throw new ArithmeticException("Division by zero")
    else left.evaluate / r
  
  def simplify: Expression =
    (left.simplify, right.simplify) match
      case (Number(0), _) => Number(0)
      case (l, Number(1)) => l
      case (Number(a), Number(b)) if b != 0 => Number(a / b)
      case (l, r) if l == r => Number(1)
      case (l, r) => Divide(l, r)
  
  def variables: Set[String] = left.variables ++ right.variables
  override def toString: String = s"($left / $right)"

object Expression:
  /** 変数に値を束縛 */
  def bind(expr: Expression, bindings: Map[String, Double]): Expression =
    expr match
      case n: Number => n
      case v: Variable => bindings.get(v.name).map(Number(_)).getOrElse(v)
      case Add(l, r) => Add(bind(l, bindings), bind(r, bindings))
      case Subtract(l, r) => Subtract(bind(l, bindings), bind(r, bindings))
      case Multiply(l, r) => Multiply(bind(l, bindings), bind(r, bindings))
      case Divide(l, r) => Divide(bind(l, bindings), bind(r, bindings))

File: apps/haskell/part3/src/CompositePattern.hs
{-# LANGUAGE RecordWildCards #-}

-- |
-- Module      : CompositePattern
-- Description : Composite pattern implementation in Haskell
-- 
-- This module demonstrates the Composite pattern using various examples:
-- - Shape: Geometric shapes with translate and scale operations
-- - Switchable: On/Off switchable devices
-- - FileSystem: Files and directories
-- - MenuItem: Menu items and set menus
-- - Expression: Mathematical expressions
-- - Organization: Organizational structure

module CompositePattern
  ( -- * Shape types and operations
    Point(..)
  , Shape(..)
  , translateShape
  , scaleShape
  , shapeArea
  , emptyComposite
  , addShape
  
  -- * Switchable types and operations
  , Switchable(..)
  , turnOn
  , turnOff
  , isOn
  , emptySwitchableComposite
  , addSwitchable
  
  -- * FileSystem types and operations
  , FileSystemEntry(..)
  , fileSize
  , fileCount
  , findByExtension
  , allFiles
  , emptyDirectory
  , addEntry
  
  -- * MenuItem types and operations
  , MenuItem(..)
  , menuPrice
  , menuItemCount
  , emptySetMenu
  , addMenuItem
  
  -- * Expression types and operations
  , Expr(..)
  , evaluate
  , simplify
  , variables
  , bindVars
  
  -- * Organization types and operations
  , OrganizationMember(..)
  , totalSalary
  , employeeCount
  , findByRole
  , emptyDepartment
  , addMember
  ) where

-- | 2D Point
data Point = Point
  { pointX :: Double
  , pointY :: Double
  } deriving (Show, Eq)

-- | Shape - Composite pattern for geometric shapes
data Shape
  = Circle Point Double           -- ^ Circle with center and radius
  | Square Point Double           -- ^ Square with top-left corner and side length
  | Rectangle Point Double Double -- ^ Rectangle with top-left, width, and height
  | CompositeShape [Shape]        -- ^ Composite containing multiple shapes
  deriving (Show, Eq)

-- | Translate a shape by (dx, dy)
translateShape :: Double -> Double -> Shape -> Shape
translateShape dx dy shape = case shape of
  Circle (Point x y) r -> Circle (Point (x + dx) (y + dy)) r
  Square (Point x y) s -> Square (Point (x + dx) (y + dy)) s
  Rectangle (Point x y) w h -> Rectangle (Point (x + dx) (y + dy)) w h
  CompositeShape shapes -> CompositeShape (map (translateShape dx dy) shapes)

-- | Scale a shape by a factor
scaleShape :: Double -> Shape -> Shape
scaleShape factor shape = case shape of
  Circle c r -> Circle c (r * factor)
  Square tl s -> Square tl (s * factor)
  Rectangle tl w h -> Rectangle tl (w * factor) (h * factor)
  CompositeShape shapes -> CompositeShape (map (scaleShape factor) shapes)

-- | Calculate the area of a shape
shapeArea :: Shape -> Double
shapeArea shape = case shape of
  Circle _ r -> pi * r * r
  Square _ s -> s * s
  Rectangle _ w h -> w * h
  CompositeShape shapes -> sum (map shapeArea shapes)

-- | Create an empty composite shape
emptyComposite :: Shape
emptyComposite = CompositeShape []

-- | Add a shape to a composite
addShape :: Shape -> Shape -> Shape
addShape s (CompositeShape shapes) = CompositeShape (shapes ++ [s])
addShape s other = CompositeShape [other, s]

-- ============================================================
-- Switchable
-- ============================================================

-- | Switchable - Composite pattern for switchable devices
data Switchable
  = Light String Bool             -- ^ Light with name and on/off state
  | DimmableLight String Int      -- ^ Dimmable light with name and intensity (0-100)
  | Fan String Bool Int           -- ^ Fan with name, on/off state, and speed
  | CompositeSwitchable String [Switchable]  -- ^ Composite with name and list
  deriving (Show, Eq)

-- | Turn on a switchable
turnOn :: Switchable -> Switchable
turnOn switchable = case switchable of
  Light name _ -> Light name True
  DimmableLight name _ -> DimmableLight name 100
  Fan name _ speed -> Fan name True (max speed 1)
  CompositeSwitchable name items -> CompositeSwitchable name (map turnOn items)

-- | Turn off a switchable
turnOff :: Switchable -> Switchable
turnOff switchable = case switchable of
  Light name _ -> Light name False
  DimmableLight name _ -> DimmableLight name 0
  Fan name _ speed -> Fan name False speed
  CompositeSwitchable name items -> CompositeSwitchable name (map turnOff items)

-- | Check if a switchable is on
isOn :: Switchable -> Bool
isOn switchable = case switchable of
  Light _ on -> on
  DimmableLight _ intensity -> intensity > 0
  Fan _ on _ -> on
  CompositeSwitchable _ items -> any isOn items

-- | Create an empty composite switchable
emptySwitchableComposite :: String -> Switchable
emptySwitchableComposite name = CompositeSwitchable name []

-- | Add a switchable to a composite
addSwitchable :: Switchable -> Switchable -> Switchable
addSwitchable item (CompositeSwitchable name items) = 
  CompositeSwitchable name (items ++ [item])
addSwitchable item other = 
  CompositeSwitchable "Group" [other, item]

-- ============================================================
-- FileSystem
-- ============================================================

-- | FileSystemEntry - Composite pattern for file system
data FileSystemEntry
  = File String Int String        -- ^ File with name, size, and extension
  | Directory String [FileSystemEntry]  -- ^ Directory with name and children
  deriving (Show, Eq)

-- | Get the total size of a file system entry
fileSize :: FileSystemEntry -> Int
fileSize entry = case entry of
  File _ size _ -> size
  Directory _ children -> sum (map fileSize children)

-- | Get the number of files in a file system entry
fileCount :: FileSystemEntry -> Int
fileCount entry = case entry of
  File {} -> 1
  Directory _ children -> sum (map fileCount children)

-- | Find files by extension
findByExtension :: String -> FileSystemEntry -> [FileSystemEntry]
findByExtension ext entry = case entry of
  File _ _ fileExt | fileExt == ext -> [entry]
  File {} -> []
  Directory _ children -> concatMap (findByExtension ext) children

-- | Get all files in a file system entry
allFiles :: FileSystemEntry -> [FileSystemEntry]
allFiles entry = case entry of
  File {} -> [entry]
  Directory _ children -> concatMap allFiles children

-- | Create an empty directory
emptyDirectory :: String -> FileSystemEntry
emptyDirectory name = Directory name []

-- | Add an entry to a directory
addEntry :: FileSystemEntry -> FileSystemEntry -> FileSystemEntry
addEntry child (Directory name children) = Directory name (children ++ [child])
addEntry child other = Directory "root" [other, child]

-- ============================================================
-- MenuItem
-- ============================================================

-- | MenuItem - Composite pattern for menu items
data MenuItem
  = Item String Double String     -- ^ Item with name, price, and category
  | SetMenu String [MenuItem] Double  -- ^ Set menu with name, items, and discount rate
  deriving (Show, Eq)

-- | Calculate the price of a menu item
menuPrice :: MenuItem -> Double
menuPrice item = case item of
  Item _ price _ -> price
  SetMenu _ items discountRate -> 
    let totalPrice = sum (map menuPrice items)
    in totalPrice * (1 - discountRate)

-- | Get the number of items in a menu item
menuItemCount :: MenuItem -> Int
menuItemCount item = case item of
  Item {} -> 1
  SetMenu _ items _ -> sum (map menuItemCount items)

-- | Create an empty set menu
emptySetMenu :: String -> Double -> MenuItem
emptySetMenu name discount = SetMenu name [] discount

-- | Add an item to a set menu
addMenuItem :: MenuItem -> MenuItem -> MenuItem
addMenuItem child (SetMenu name items discount) = SetMenu name (items ++ [child]) discount
addMenuItem child other = SetMenu "Set" [other, child] 0.0

-- ============================================================
-- Expression
-- ============================================================

-- | Expression - Composite pattern for mathematical expressions
data Expr
  = Number Double                 -- ^ Numeric literal
  | Variable String (Maybe Double)  -- ^ Variable with optional bound value
  | Add Expr Expr                 -- ^ Addition
  | Subtract Expr Expr            -- ^ Subtraction
  | Multiply Expr Expr            -- ^ Multiplication
  | Divide Expr Expr              -- ^ Division
  | Negate Expr                   -- ^ Negation
  deriving (Show, Eq)

-- | Evaluate an expression
evaluate :: Expr -> Maybe Double
evaluate expr = case expr of
  Number n -> Just n
  Variable _ (Just v) -> Just v
  Variable name Nothing -> Nothing  -- Unbound variable
  Add l r -> (+) <$> evaluate l <*> evaluate r
  Subtract l r -> (-) <$> evaluate l <*> evaluate r
  Multiply l r -> (*) <$> evaluate l <*> evaluate r
  Divide l r -> do
    lv <- evaluate l
    rv <- evaluate r
    if rv == 0 then Nothing else Just (lv / rv)
  Negate e -> negate <$> evaluate e

-- | Simplify an expression
simplify :: Expr -> Expr
simplify expr = case expr of
  Add l r -> case (simplify l, simplify r) of
    (Number 0, r') -> r'
    (l', Number 0) -> l'
    (Number a, Number b) -> Number (a + b)
    (l', r') -> Add l' r'
  Subtract l r -> case (simplify l, simplify r) of
    (l', Number 0) -> l'
    (Number a, Number b) -> Number (a - b)
    (l', r') | l' == r' -> Number 0
    (l', r') -> Subtract l' r'
  Multiply l r -> case (simplify l, simplify r) of
    (Number 0, _) -> Number 0
    (_, Number 0) -> Number 0
    (Number 1, r') -> r'
    (l', Number 1) -> l'
    (Number a, Number b) -> Number (a * b)
    (l', r') -> Multiply l' r'
  Divide l r -> case (simplify l, simplify r) of
    (Number 0, _) -> Number 0
    (l', Number 1) -> l'
    (Number a, Number b) | b /= 0 -> Number (a / b)
    (l', r') -> Divide l' r'
  Negate e -> case simplify e of
    Number n -> Number (-n)
    Negate e' -> e'  -- Double negation
    e' -> Negate e'
  e -> e

-- | Get all variable names in an expression
variables :: Expr -> [String]
variables expr = case expr of
  Number _ -> []
  Variable name _ -> [name]
  Add l r -> variables l ++ variables r
  Subtract l r -> variables l ++ variables r
  Multiply l r -> variables l ++ variables r
  Divide l r -> variables l ++ variables r
  Negate e -> variables e

-- | Bind values to variables
bindVars :: [(String, Double)] -> Expr -> Expr
bindVars bindings expr = case expr of
  Variable name Nothing -> 
    case lookup name bindings of
      Just v -> Variable name (Just v)
      Nothing -> expr
  Variable {} -> expr
  Number n -> Number n
  Add l r -> Add (bindVars bindings l) (bindVars bindings r)
  Subtract l r -> Subtract (bindVars bindings l) (bindVars bindings r)
  Multiply l r -> Multiply (bindVars bindings l) (bindVars bindings r)
  Divide l r -> Divide (bindVars bindings l) (bindVars bindings r)
  Negate e -> Negate (bindVars bindings e)

-- ============================================================
-- Organization
-- ============================================================

-- | OrganizationMember - Composite pattern for organizational structure
data OrganizationMember
  = Employee String Double String  -- ^ Employee with name, salary, and role
  | Department String (Maybe String) [OrganizationMember]  -- ^ Department with name, manager, and members
  deriving (Show, Eq)

-- | Calculate the total salary
totalSalary :: OrganizationMember -> Double
totalSalary member = case member of
  Employee _ salary _ -> salary
  Department _ _ members -> sum (map totalSalary members)

-- | Get the number of employees
employeeCount :: OrganizationMember -> Int
employeeCount member = case member of
  Employee {} -> 1
  Department _ _ members -> sum (map employeeCount members)

-- | Find employees by role
findByRole :: String -> OrganizationMember -> [OrganizationMember]
findByRole role member = case member of
  Employee _ _ r | r == role -> [member]
  Employee {} -> []
  Department _ _ members -> concatMap (findByRole role) members

-- | Create an empty department
emptyDepartment :: String -> Maybe String -> OrganizationMember
emptyDepartment name manager = Department name manager []

-- | Add a member to a department
addMember :: OrganizationMember -> OrganizationMember -> OrganizationMember
addMember child (Department name manager members) = Department name manager (members ++ [child])
addMember child other = Department "Organization" Nothing [other, child]
